resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesConf.scala

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.k8s
+
+import org.apache.spark.{SparkConf, SparkException}
+import org.apache.spark.deploy.k8s.Config._
+import org.apache.spark.deploy.k8s.Constants._
+import org.apache.spark.deploy.k8s.submit.MainAppResource
+import org.apache.spark.internal.config.ConfigEntry
+
+private[k8s] sealed trait KubernetesRoleSpecificConf
+private[k8s] case class KubernetesDriverSpecificConf(
+  private val sparkConf: SparkConf,
+  mainAppResource: Option[MainAppResource],
+  mainClass: String,
+  appName: String,
+  appArgs: Seq[String]) extends KubernetesRoleSpecificConf {
+  def driverAnnotations(): Map[String, String] = {
+    val driverCustomAnnotations = KubernetesUtils.parsePrefixedKeyValuePairs(
+      sparkConf, KUBERNETES_DRIVER_ANNOTATION_PREFIX)
+    require(!driverCustomAnnotations.contains(SPARK_APP_NAME_ANNOTATION),
+      s"Annotation with key $SPARK_APP_NAME_ANNOTATION is not allowed as it is reserved for" +
+        " Spark bookkeeping operations.")
+    driverCustomAnnotations ++ Map(SPARK_APP_NAME_ANNOTATION -> appName)
+  }
+}
+
+private[k8s] case object KubernetesExecutorSpecificConf extends KubernetesRoleSpecificConf
+
+private[k8s] class KubernetesConf[T <: KubernetesRoleSpecificConf](
+    private val sparkConf: SparkConf,
+    val roleSpecificConf: T,
+    val appResourceNamePrefix: String,
+    val appId: String) {
+
+  def kubernetesAppId(): String = sparkConf
+    .getOption("spark.app.id")
+    .getOrElse(throw new SparkException("Invalid config state; an app id should always" +
+      " be present."))
+
+  def namespace(): String = sparkConf.get(KUBERNETES_NAMESPACE)
+
+  def sparkJars(): Seq[String] = sparkConf
+    .getOption("spark.jars")
+    .map(str => str.split(",").toSeq)
+    .getOrElse(Seq.empty[String])
```

Comment:
`additionalMainAppJar`?

Reply:
Plan is to have the caller inject the main app resource into the `spark.jars` field.

Reply:
Change of plans - see `KubernetesConf.createDriverConf`.